### DISCUSSION TITLE
dynamic fetch on getServerSideProps??

### DISCUSSION THREAD

**Discussion by devalpha18** [vercel/next.js#38179] [Help] [answered]
Hi,
I am planning to fetch data with dynamic api(next api) on getServerSideProps
I could get dynamic route id like this
```bash
import { useRouter } from "next/router

...
const router = useRouter()
const { id } = router.query;
// id is dynamic
...
```
How can I get this on getServerSideProps ? Thanks in advance!

```bash
export const getServerSideProps = () => {
// here
}
```

**Comment by tofuwu-dev** [comment, marked as answer]
Hi, @moriyamahideki 
It is easy and it is very similar.
You can write the code as is:
```typescript
export default const Demo = ({ result }) => {
# use result as props.
}
# gSSP
export const getServerSideProps= async (context) => {
  const { id } = context.query;
  # call next api dynamically each request
  const res = await fetch(`/api/${id}`);
  const result = await res.json();

  return {
    props: {
      result,
    },
  };
};
```

**Reply by tofuwu-dev** [reply]
fyi: plz read this doc https://nextjs.org/docs/api-reference/data-fetching/get-server-side-props#context-parameter

**Reply by icyJoseph** [reply]
Hi,

This (reaching out for API routes from GSSP) is not recommended. GetServerSideProps is already on the server, so you shouldn't reach out for `api/routes` from within it.

https://nextjs.org/docs/basic-features/data-fetching/get-server-side-props#getserversideprops-or-api-routes:

> It can be tempting to reach for an [API Route](https://nextjs.org/docs/api-routes/introduction) when you want to fetch data from the server, then call that API route from getServerSideProps. This is an unnecessary and inefficient approach, as it will cause an extra request to be made due to both getServerSideProps and API Routes running on the server.

Just connect to the service you want to connect to directly.

**Reply by devalpha18** [reply]
It worked on my side, that is what I needed. thanks @haodev007

**Reply by devalpha18** [reply]
@icyJoseph 
I wanted to call dynamic route which already made...
It will be used several times for several purposes. so I wanted to use next api which I already made, on getServerSideProps..
How can I do it for this case?

**Reply by Anoosha0-Shetty** [reply]
Hello in my case, the folder structure is under pages for Nextjs 14
![image](https://github.com/vercel/next.js/assets/17610968/d0c36361-196a-4715-9ca5-228ce59dab73)

example, how do i use the same code for '/' & '/cart'

**Comment by icyJoseph** [comment]
```tsx
import type { GetServerSidePropsContext, NextPage } from "next";
import Link from "next/link";

const Home: NextPage = () => {
  return (
    <div>
      Home <Link href="/blog">blog</Link>
    </div>
  );
};

export const getServerSideProps = (ctx: GetServerSidePropsContext) => {
  /**
   * If this page uses a dynamic route, params contains the route parameters.
   *
   * - If the page name is [id].js , then params will look like { id: ... }
   *    - /posts/[id] -> /posts/123 then params is { id: '123' }
   *
   * - If the page name is [...id].s, then params will look like { id: [...] }
   *    - /posts/[...id] -> /posts/123 then params is { id: ['123'] }
   */
  console.log(ctx.params);
  return { props: {} };
};

export default Home;
```

**Reply by icyJoseph** [reply]
@moriyamahideki 

When you use `api/routes` from `GSSP` (GetServerSideProps), you are sort of using networking to solve a problem that could be solved by using a proper folder structure.

`GSSP` is a kind of `api/route`, that's specialised to work with Nextjs' React pages. In that sense, most things `api/routes` can do, `GSSP` can also do.

For example, fetching data from 3rd party services. At the end of the day, the contents of `api/routes` can be modelled as function that given som inputs, resolve to some output, so we could refactor this contents, into a function that we could invoke from within `api/routes` and also from `GSSP`. This is the folder structure/code architecture in action.

This way, you could even re-use, with some pre-conditions, this function from the client! Because `fetch` is poly-filled for server and client by Next.js

### Example

Define the function that gets data

```ts
// lib/codewars.ts
export const getCodeWarsUser = (user) =>
  axios
    .get(`https://www.codewars.com/api/v1/users/${user}`)
    .then(({ data }) => data);
```

Then `GSSP`

```tsx
import { getCodeWarsUser } from 'lib/codewars';

export async function getServerSideProps(context: GetServerSidePropsContext) {
  // calculate user from context
  const codewars = await getCodeWarsUser("some-user");

  return {
    props: { codewars },
  };
}
```

or in `api/routes`

```ts
// api/codewars.ts
import { getCodeWarsUser } from 'lib/codewars';

const codewars = async (req, res) => {
 // calculate "some-user" from req
  const data = await getCodeWarsUser('some-user');

  return res.send(data);
};

export default codewars;
```

### Bundle sizes and Code elimination

Code that's only used in `GSSP` is not included in the client bundle, including secrets and others. Use this tool to verify that behaviour, called code-elimination, https://next-code-elimination.vercel.app/.
